File: src/main/java/org/cimsbioko/search/IndexingService.java
package org.cimsbioko.search;

import android.content.Context;
import android.content.Intent;
import android.util.Log;
import androidx.annotation.NonNull;
import androidx.core.app.JobIntentService;

import java.io.IOException;

public class IndexingService extends JobIntentService {

    private static final String TAG = IndexingService.class.getSimpleName();

    private static final String ENTITY_TYPE = "entityType";
    private static final String ENTITY_UUID = "entityUuid";

    private static final int JOB_ID = 0xFB;

    public static void enqueueWork(Context context, Intent intent) {
        enqueueWork(context, IndexingService.class, JOB_ID, intent);
    }

    public enum EntityType {
        HIERARCHY,
        LOCATION,
        INDIVIDUAL
    }

    @Override
    protected void onHandleWork(@NonNull Intent intent) {
        Indexer indexer = Indexer.getInstance();
        if (intent.hasExtra(ENTITY_UUID)) {
            EntityType type = EntityType.valueOf(intent.getStringExtra(ENTITY_TYPE));
            String uuid = intent.getStringExtra(ENTITY_UUID);
            try {
                switch (type) {
                    case HIERARCHY:
                        indexer.reindexHierarchy(uuid);
                        break;
                    case LOCATION:
                        indexer.reindexLocation(uuid);
                        break;
                    case INDIVIDUAL:
                        indexer.reindexIndividual(uuid);
                        break;
                    default:
                        Log.w(TAG, "unknown entity type " + type);
                }
            } catch (IOException e) {
                Log.e(TAG, "failed during reindex", e);
            }
        } else {
            indexer.reindexAll();
        }
    }

    public static void queueFullReindex(Context ctx) {
        IndexingService.enqueueWork(ctx.getApplicationContext(), new Intent(ctx, IndexingService.class));
    }

    public static void queueReindex(Context ctx, EntityType type, String uuid) {
        Intent intent = new Intent(ctx, IndexingService.class);
        intent.putExtra(ENTITY_TYPE, type.toString());
        intent.putExtra(ENTITY_UUID, uuid);
        IndexingService.enqueueWork(ctx.getApplicationContext(), intent);
    }
}

File: src/main/java/org/cimsbioko/search/IndexingService.kt
package org.cimsbioko.search

import android.content.Context
import android.content.Intent
import android.util.Log
import androidx.core.app.JobIntentService
import java.io.IOException

class IndexingService : JobIntentService() {

    enum class EntityType {
        HIERARCHY, LOCATION, INDIVIDUAL
    }

    override fun onHandleWork(intent: Intent) {
        with(Indexer.instance) {
            if (intent.hasExtra(ENTITY_UUID)) {
                val type = EntityType.valueOf(intent.getStringExtra(ENTITY_TYPE))
                val uuid = intent.getStringExtra(ENTITY_UUID)
                try {
                    when (type) {
                        EntityType.HIERARCHY -> reindexHierarchy(uuid)
                        EntityType.LOCATION -> reindexLocation(uuid)
                        EntityType.INDIVIDUAL -> reindexIndividual(uuid)
                    }
                } catch (e: IOException) {
                    Log.e(TAG, "failed during reindex", e)
                }
            } else {
                reindexAll()
            }
        }
    }

    companion object {

        private val TAG = IndexingService::class.java.simpleName
        private const val ENTITY_TYPE = "entityType"
        private const val ENTITY_UUID = "entityUuid"
        private const val JOB_ID = 0xFB

        private fun enqueueWork(context: Context, intent: Intent) {
            enqueueWork(context, IndexingService::class.java, JOB_ID, intent)
        }

        @JvmStatic
        fun queueFullReindex(ctx: Context) {
            enqueueWork(ctx.applicationContext, Intent(ctx, IndexingService::class.java))
        }

        @JvmStatic
        fun queueReindex(ctx: Context, type: EntityType, uuid: String) {
            enqueueWork(ctx.applicationContext, Intent(ctx, IndexingService::class.java).apply {
                putExtra(ENTITY_TYPE, type.toString())
                putExtra(ENTITY_UUID, uuid)
            })
        }
    }
}